Action RPG/Player/Player.gd

```diff
@@ -56,10 +56,10 @@ func move_state(delta):
 		animationTree.set("parameters/Attack/blend_position", input_vector)
 		animationTree.set("parameters/Roll/blend_position", input_vector)
 		animationState.travel("Run")
-		velocity = velocity.move_toward(input_vector * MAX_SPEED, ACCELERATION * delta)
```

Comment:
It makes acceleration and friction frame-dependent, while velocity will be time dependent. It can cause some slow accelerations when fps drops. Probably should just tolerate high values for these

Reply:
Only velocity is used to change the players' position, which is already multiplied by delta inside move_and_slide()

Because the velocity is always updated starting the Vector2.ZERO, framerate has no effect on it, and the new velocity is being calculated using absolute values of MAX_SPEED, ACCELERATION, and FRICTION.

The values that I suggested in the PR kinda pass the same feel when I tested before/after. But I don't have different setups to test it further. Anyways, I do think using delta here is unnecessary.

Reply:
Velocity is the change in position, which like you said, doesn't need to be multiplied by delta because move_and_slide already does that

However, friction and acceleration are the change in velocity. So, if you don't multiply them by delta the change in velocity will be frame dependent (you may have acceleration and deceleration stutters). Artelhin is correct.

Reply:
Please, let me know if I'm just being annoying here, I just want to make sure my point is clear.

Acceleration/Friction is nothing more than adding/subtracting a value from current velocity

Taking the code from the MR as example:
```
velocity = velocity.move_toward(input_vector * MAX_SPEED, ACCELERATION)
velocity = velocity.move_toward(Vector2.ZERO, FRICTION)
velocity = move_and_slide(velocity)
```
We can view acceleration simply as `v1 = v0 + a` 
_v0_ is the current velocity, which is already corrected by delta (from previous update)
_a_ is the acceleraction we want

then we could do the multiplication by delta manually:
`v1 = v0 * delta + a * delta`

but because _delta_ here is the common factor for both _v0_ and _a_, you can put it in evidence

`v1 = (v0 + a) * delta`  and this is what _move_and_slide_ is doing

the original code is then doing:
`v1 = (v0 + a * delta) * delta` and this is why you need 500 as acceleration, to compensate the double multiplication by delta

We can do a simple experiment using these values as example:
```
MAX_SPEED = 80
A1 = 500
A2 = 16
```

`v1 = (v0 + A1 * delta) * delta`
or
`v1 = (v0 + A2) * delta`
It will take 5 updates to reach MAX_SPEED.
I tested with different FPS values 15, 30, 60, and unlimited (by changing in project settings)

```
(0, 0)
(16.666668, 0)
(33.333336, 0)
(50.000004, 0)
(66.666672, 0)
(80, 0)
(64, 0)
(48, 0)
(32, 0)
(16, 0)
(0, 0)
```
Here is the log from the test, using ACCELERATION with delta and FRICTION without.

The only difference is having more clarity and a cleaner code by not using delta directly in the update.
You can have a better control over the velocity, with values that makes sense to the player (if you want to show it).

Again, it's just a matter of clarity, and you can always have a workaround to make things work for you, but I prefer simpler solutions.

Reply:
The point is that acceleration is the change in speed for a given period of time (usually 1 second in mechanical physics)
So when we multiply it by delta, we apply only a part of a whole acceleration value, so when we add up all the deltas to some period of time, say, a second, the sum of speed differences we get is the same, whenever we measure this sum.

When you don't multiply by delta, you apply the same acceleration every physics frame, so the change in speed in a second starts to depend on a number of frames, which happened during it.
Usually it's not a problem, because the number of frames per second is mostly consistent.

The idea of basing acceleration and speed on the number of frames is not wrong in itself, but if you base acceleration on fps, then you should base the speed on the fps too (not multiply velocity by delta), otherwise it's a mix between two different ideas, which are not meant to be mixed.
However, move_and_slide() multiplies velocity by delta, so we don't really have a choice here.

Reply:
We both agree that time-based movement is usually the best approach in these cases, but I'm not seeing how my PR violates that.

The value of velocity that is passed to move_and_slide already includes the aceleration for that update, and then the function does the job of multiplying it by delta. Maybe I'm missing something.

Reply:
I suggest considering the formulas for movements from physics, maybe:
When we start from v=0
v(t) = a*t
If we start from some speed u:
v(t) = u + a*t
The last one is what we have in pur process function

Let's also consider these cases (acceleration=50):
1s time, 4 fps, delta-based acceleration  - final speed will be 50*1/4+50*1/4+50*1/4+50*1/4=50
1s time, 2fps, delta-based acceleration - final speed will be 50*1/2+50*1/2=50
1s time, 4 fps, no-delta acceleration - final speed will be 50+50+50+50=200
1s time, 2 fps, no-delta acceleration - final speed will be 50+50=100

As you can see, when we don't multiply by delta, we get our final speed wrong, when fps drops to lower values.
However, in time-based movement it is expected to be the same for any given period of time, no matter the fps.
This is where your PR gets it wrong. Hope this was of help.

Reply:
> 

No worries, I don't mind seeking some clarification.

Your example here prints out outputs for each frame. It will of course take the same amount frames to reach the max speed/min speed (as your prints show). However, what your example fails to show is that it would take a different amount of time in the real world because your friction doesn't use delta.

As an example, image your game is running at 60 fps and your acceleration is 1 unit per frame. That means you'd reach a velocity of 60 units in 1 second.

However, if your game is running at 30 fps and your acceleration is still 1 unit per frame (because you aren't multiplying by delta) you will reach a velocity of 60 units in 2 seconds.

Hopefully that makes sense.

Reply:
acceleration was added to the current velocity in the line:
`velocity = velocity.move_toward(input_vector * MAX_SPEED, ACCELERATION)`

which is basically `v1 = v0 + a`

then the updated velocity is passed to _move_and_slide_

Reply:
that works until you change the speed that the physics updates. which is important if something prevents your game from running at the default 60 physics frames a second(even if its brief). To test this go to your project settings and change your `Physics>Common>Physics Fps` to say... 2.... if you did things correctly your character should traverse the same distance in the same amount of real world time despite updating at a much lower framerate.

Reply:
@rhrlima I agree with you on the clarity problem, but I won't accept a solution that makes acceleration or friction frame dependent.

Did the frame rate example in my last comment make sense to you?

Reply:
It does make sense @uheartbeast . You can close the PR if you want, I won't complain.

I'll just leave this last contribution:
```
extends KinematicBody2D

var velocity = Vector2.ZERO
var MAX_SPEED = 200
var ACCELERATION = 50

func _ready():
	$Timer.start()

func _physics_process(delta):
	velocity = velocity.move_toward(Vector2(1, 0) * MAX_SPEED, ACCELERATION)
	velocity = move_and_slide(velocity)

func _on_Timer_timeout():
	MAX_SPEED = 0
```

I did a small test to check different values for Physics FPS, as mentioned by @Salvakiya 

The code is basically a kinematic body, a sprite and a timer that runs for 2 seconds.
I tested using 2/30/60/120 physics FPS.

Besides using 2 FPS, for each test the sprite moved the same amount of pixels. For 2 FPS, the difference was 7 pixels more than the others.

Reply:
However given that example it only takes 4 physics frames for velocity to get to (200,0). but it does cap at that value. this is why it can seem to be correct. The change in velocity should be multiplied by the delta. Or at least it would be ill advised to do otherwise as it could complicate things further on in the games development.